google/detectors/credentials/generic_weak_credential_detector/src/main/java/com/google/tsunami/plugins/detectors/credentials/genericweakcredentialdetector/testers/hsqldb/HyperSQLCredentialTester.java

```diff
@@ -0,0 +1,110 @@
+import com.google.tsunami.plugins.detectors.credentials.genericweakcredentialdetector.provider.TestCredential;
+import com.google.tsunami.plugins.detectors.credentials.genericweakcredentialdetector.tester.CredentialTester;
+import com.google.tsunami.proto.NetworkService;
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.util.List;
+import javax.inject.Inject;
+
+/** Credential tester specifically for HyperSQL. */
+public final class HyperSQLCredentialTester extends CredentialTester {
+  private static final GoogleLogger logger = GoogleLogger.forEnclosingClass();
+  private final ConnectionProviderInterface connectionProvider;
+
+  private static final ImmutableMap<String, TargetService> SERVICE_MAP =
+      ImmutableMap.of("jdbc", TargetService.HSQLDB);
+
+  @Inject
+  HyperSQLCredentialTester(ConnectionProviderInterface connectionProvider) {
+    this.connectionProvider = checkNotNull(connectionProvider);
+  }
+
+  @Override
+  public String name() {
+    return "HyperSQLCredentialTester";
+  }
+
+  @Override
+  public String description() {
+    return "HyperSQL credential tester.";
+  }
+
+  @Override
+  public boolean canAccept(NetworkService networkService) {
+    String serviceName = NetworkServiceUtils.getServiceName(networkService);
+    return SERVICE_MAP.containsKey(serviceName);
+  }
+
+  @Override
+  public boolean batched() {
+    return true;
+  }
+
+  @Override
+  public ImmutableList<TestCredential> testValidCredentials(
+      NetworkService networkService, List<TestCredential> credentials) {
+    if (!canAccept(networkService)) {
+      return ImmutableList.of();
+    }
+
+    return credentials.stream()
+        .filter(cred -> isHsqlAccessible(networkService, cred))
+        .collect(toImmutableList());
+  }
+
+  /**
+   * Using testdb as database name since hsqldb requires a database name in order to perform the
+   * connection However hsqldb does not create a default database during installation testdb is the
+   * database name used within the documentation
+   */
+  private boolean isHsqlAccessible(NetworkService networkService, TestCredential credential) {
+    try {
+      var url =
+          String.format(
+              "jdbc:hsqldb:hsql://%s/testdb",
+              NetworkEndpointUtils.toUriAuthority(networkService.getNetworkEndpoint()));
+      logger.atInfo().log(
+          "url: %s, username: %s, password: %s",
+          url, credential.username(), credential.password().orElse(""));
+      Connection conn =
+          connectionProvider.getConnection(
+              url, credential.username(), credential.password().orElse(""));
+
+      if (conn != null) {
+        logger.atInfo().log("Connected to the Hyper SQL server successfully.");
+        return true;
+      }
+    } catch (SQLException e) {
+      logger.atSevere().log(
+          "HyperSQLCredentialTester sql error: %s (%d)", e.getMessage(), e.getErrorCode());
+    }
```

Comment:
Error code `-4001` is returned when the user is invalid.  
Error code `-4000` is returned when the user exists but the password is invalid.  

This check optimizes testing by an order of magnitude, as it tries common passwords only on valid usernames.

```suggestion
    } catch (SQLException e) {
      if (e.getErrorCode() == HSQL_INVALID_USER_ERROR) {
        invalidUsernames.add(credential.username());
      }

      logger.atSevere().log(
          "HyperSQLCredentialTester sql error: %s (%d)", e.getMessage(), e.getErrorCode());
    }
```